src/main/java/io/vertx/core/http/impl/HttpServerImpl.java

```diff
@@ -678,7 +711,7 @@ protected void handleMessage(ServerConnection conn, ContextImpl context, Channel
             if (!closeFrameSent) {
               // Echo back close frame and close the connection once it was written.
               // This is specified in the WebSockets RFC 6455 Section  5.4.1
-              ch.writeAndFlush(wsFrame).addListener(ChannelFutureListener.CLOSE);
+              ch.writeAndFlush(new CloseWebSocketFrame().replace(wsFrame.getBinaryData())).addListener(ChannelFutureListener.CLOSE);
```

Comment:
can you explain this change ?

Reply:
It was assembling the PONG message empty. An empty PONG message doesn't get sent, the client doesn't see it and disconnects. This puts the content in the PONG.

Reply:
ok, sure but then why not do instead:

`ch.writeAndFlush(new CloseWebSocketFrame(true, 0, wsFrame.getBinaryData())).addListener(ChannelFutureListener.CLOSE);`

Reply:
this is a reply to a CLOSE msg not a PONG here

Reply:
Sorry, my mistake. I made that change at the same time as the PONG fix. I think it was a similar issue. There's some code further down the stack in netty (handleWebSocketFrame I think) that was silently dropping the echo of both the pong and close because they were not of one of the WebSocketFrame child classes.